File: 3518-maximum-multiplication-score/3518-maximum-multiplication-score.cs
public class Solution {
    long Dfs(int[] a, int[] b, int i, int j, long[][] dp)
    {
        if (i == 4) return 0;
        if (j == b.Length) return int.MinValue;
        if (dp[i][j] != -1) return dp[i][j];
        long take = (long)a[i] * (long)b[j] + Dfs(a, b, i+1, j+1, dp);
        long ntake = Dfs(a, b, i, j+1, dp);
        return dp[i][j] = Math.Max(take, ntake);
    }
    public long MaxScore(int[] a, int[] b) 
    {
        long[][] dp = new long[4][];
        for (int i = 0; i < 4; i++) {
            dp[i] = new long[b.Length+1];
            for (int j = 0; j < b.Length; j++) {
                dp[i][j] = -1;
            }
        }
        return Dfs(a, b, 0, 0, dp);
    }
}

File: 3518-maximum-multiplication-score/3518-maximum-multiplication-score.cpp
class Solution {
public:
    long long dfs(vector<int>&a, vector<int>&b, int i, int j, 
    vector<vector<long long>>&dp)
    {
        if (i == 4) return 0;
        if (j == b.size())  return -1e9;
        if (dp[i][j] != -1) return dp[i][j];
        long long take = (long long)a[i]*(long long)b[j] + dfs(a, b, i+1, j+1, dp);
        long long ntake = dfs(a, b, i, j+1, dp);
        return dp[i][j] = max(take, ntake);
    }
    long long maxScore(vector<int>& a, vector<int>& b) 
    {
        vector<vector<long long>> dp(4, vector<long long>(b.size()+1, -1));
        return dfs(a, b, 0, 0, dp);
    }
};